include/ua_securitycontext.h

```diff
@@ -0,0 +1,196 @@
+                                          const UA_ByteString *const iv,
+                                          void *const contextData);
+    /**
+     * \brief Parses a given certificate to extract the remote public key from it.
+     * Fails, if the certificate is invalid or expired.
+     *
+     * \param securityPolicy contains the function pointers associated with the policy.
+     * \param remoteCertificate the remote certificate to extract the client public key from.
+     * \param contextData the context to work on.
+     */
+    UA_StatusCode (*const parseRemoteCertificate)(const UA_SecurityPolicy *const securityPolicy,
+                                                  const UA_ByteString *const remoteCertificate,
+                                                  void *const contextData);
+
+    /**
+     * \brief Gets the signature size that depends on the remote public key.
+     *
+     * \param securityPolicy contains the function pointers associated with the policy.
+     * \param contextData the context to retrieve data from.
+     * \return the size of the remote asymmetric signature. Returns 0 if no remote certificate
+     *                     was set previousely.
+     */
+    size_t (*const getRemoteAsymSignatureSize)(const UA_SecurityPolicy *const securityPolicy,
+                                               const void *const contextData);
+
+    /**
+     * \brief Gets the plaintext block size that depends on the remote public key.
+     *
+     * \param securityPolicy contains the function pointers associated with the policy.
+     * \param contextData the context to retrieve data from.
+     * \return the size of the plain text block size when encrypting with the remote public key.
+     *         Returns 0 as long as no remote certificate was set previousely.
+     */
+    size_t (*const getRemoteAsymPlainTextBlockSize)(const UA_SecurityPolicy *const securityPolicy,
+                                                    const void *const contextData);
+
+    /**
+     * Gets the number of bytes that are needed by the encryption function in addition to the length of the plaintext message.
+     * This is needed, since most rsa encryption methods have their own padding mechanism included. This makes the encrypted
+     * message larger than the plainText, so we need to have enough room in the buffer for the overhead.
+     *
+     * \param securityPolicy contains the function pointers associated with the policy.
+     * \param contextData the retrieve data from.
+     * \param maxEncryptionLength the maximum number of bytes that the data to encrypt can be.
+     */
+    size_t (*const getRemoteAsymEncryptionBufferLengthOverhead)(const UA_SecurityPolicy *const securityPolicy,
+                                                                const void *const contextData,
+                                                                const size_t maxEncryptionLength);

```

Comment:
Again, the caller does not care if maxEncryptionLength is const, as the argument is passed-by-value.